elsirion/webimint-rs
# Save Successfully-connected Federations to Local Storage

## Why
It's difficult to have to paste in a fedimint invite code every time the page reloads.  Additionally, if you enter an invalid federation there's just a blank screen and no error indications.

## Changes
1. Added a "Try Again" screen (`src/components/failure.rs`) if the federation client fails to join the specified invite code. Pressing the "Try again" button refreshes the page.
  ![Screenshot 2023-09-19 at 12 44 32 PM](https://github.com/elsirion/webimint-rs/assets/50180265/3ed7e0b9-7f25-4ac5-948f-a507571f3128)
2. Save successfully-connected federations to Local Storage. Both the federation invite code (`src/components/app.rs:27`) and the federation name/label (`src/components/joined.rs:46`) are saved to local storage to be accessed later. If the localstorage values exist, a clickable component will be shown on the initial connection screen.  Clicking it auto-fills the input.
  ![Screenshot 2023-09-19 at 12 48 42 PM](https://github.com/elsirion/webimint-rs/assets/50180265/d7343844-e669-4c51-aad4-ce862e5c9706)

### Other Changes
 - `Cargo.toml` - added the `Storage` feature to `web-sys` to enable the use of `window.localStorage`
 - `src/utils/local_storage.rs` - created a `local_storage` util that returns `window.localStorage` over `web-sys`

## Test Plan
1. Ensure fedimint-cli is running
2. Get an invite code in fedimint-cli by running `cat target/devimint/cfg/invite-code` in the fedimint shell
3. Run webimint-rs with `trunk serve`
4. Join the federation via invite code
5. Refresh the page
6. You will see the "Join an existing federation" component
7. Clicking the saved federation should autofill the form
8. Refresh the page
9. Enter an invalid invite code (any string)
10. The "Try again" screen will appear
11. Clicking the "Try again" button will refresh the page
